clusterloader2/pkg/measurement/common/ooms_tracker.go

```diff
@@ -0,0 +1,216 @@
+	case "start":
+		if err = m.start(config.ClusterFramework.GetClientSets().GetClient()); err != nil {
+			return nil, fmt.Errorf("starting cluster OOMs measurement problem: %w", err)
+		}
+		return nil, nil
+	case "gather":
+		m.oomsLock.Lock()
+		defer m.oomsLock.Unlock()
+		summary, err := m.gather()
+		if err != nil {
+			return nil, fmt.Errorf("gathering cluster OOMs measurement problem: %w", err)
+		}
+		return summary, m.validate()
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (m *clusterOomsTrackerMeasurement) Dispose() {
+	m.stop()
+}
+
+func (m *clusterOomsTrackerMeasurement) String() string {
+	return clusterOomsTrackerName
+}
+
+func (m *clusterOomsTrackerMeasurement) start(c clientset.Interface) error {
+	if m.isRunning {
+		klog.Infof("%s: cluster OOMs tracking measurement already running", m)
+		return nil
+	}
+	klog.Infof("%s: starting cluster OOMs tracking measurement...", m)
+	m.initFields()
+	// Watching for OOM events from node-problem-detector below.
+	i := informer.NewInformer(
+		c,
+		"events",
+		m.selector,
+		m.handleOomEvent,
+	)
+	if err := informer.StartAndSync(i, m.stopCh, informerTimeout); err != nil {
+		return fmt.Errorf("problem with OOM events informer starting: %w", err)
+	}
+	return nil
+}
+
+func (m *clusterOomsTrackerMeasurement) initFields() {
+	m.isRunning = true
+	m.stopCh = make(chan struct{})
+	m.ooms = make([]oomEvent, 0)
+	m.selector = &measurementutil.ObjectSelector{
+		FieldSelector: fields.Set{"reason": oomEventReason}.AsSelector().String(),
+		Namespace:     metav1.NamespaceAll,
+	}
+	m.msgRegex = oomEventMsgRegex
+}
+
+func (m *clusterOomsTrackerMeasurement) stop() {
+	if m.isRunning {
+		m.isRunning = false
+		close(m.stopCh)
+	}
+}
+
+func (m *clusterOomsTrackerMeasurement) gather() ([]measurement.Summary, error) {
+	klog.Infof("%s: gathering cluster OOMs tracking measurement", clusterOomsTrackerName)
+	if !m.isRunning {
+		return nil, fmt.Errorf("measurement %s has not been started", clusterOomsTrackerName)
+	}
+
+	m.stop()
+
+	content, err := util.PrettyPrintJSON(struct {
+		Ooms []oomEvent `json:"ooms"`
+	}{
+		Ooms: m.ooms,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("OOMs PrettyPrintJSON problem: %w", err)
+	}
+
+	summary := measurement.CreateSummary(clusterOomsTrackerName, "json", content)
+	return []measurement.Summary{summary}, nil
+}
+
+func (m *clusterOomsTrackerMeasurement) validate() error {
+	if len(m.ooms) == 0 {
+		return nil
+	}
+	return fmt.Errorf("OOMs recorded: %+v", m.ooms)
+}
+
+func (m *clusterOomsTrackerMeasurement) handleOomEvent(_, obj interface{}) {
+	if obj == nil {
+		return
+	}
+	event, ok := obj.(*corev1.Event)
+	if !ok {
+		return
+	}
+
+	oom := oomEvent{
+		Node: event.InvolvedObject.Name,
+		Time: event.FirstTimestamp.Time,
+	}
+
+	match := m.msgRegex.FindStringSubmatch(event.Message)
+	if len(match) != 4 {
+		klog.Warningf(`unexpected OOM event message regex mismatch; event message contents: "%v"`, event.Message)
```

Comment:
should we do something else?
Tht warning may suggest that OOM events have a new format.

Maybe the better thing to do is to basically report an OOM with not all fields being set to present the problem?